### PR TITLE
Refactoring UDMIS foundation and adding overview doc

### PR DESCRIPTION
**Pull request by grafnu**
The [rendered documentation](https://grafnu.github.io/udmi/docs/udmis/class_overview.html) shows what it looks like in graphical form.

### CODE REVIEW

File: udmis/src/main/java/com/google/bos/udmi/service/messaging/impl/MessageDispatcherImpl.java [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+  @TestOnly
+  public void resetForTest() {
+    ((MessageBase) messagePipe).resetForTest();
+  }
+
+  /**
+   * Process a received message bundle.
+   */
+  void processMessage(Bundle bundle) {
+    Envelope envelope = Preconditions.checkNotNull(bundle.envelope, "bundle envelope is null");
```
Comment (johnrandolph) [inline udmis/src/main/java/com/google/bos/udmi/service/messaging/impl/MessageDispatcherImpl.java:180]: Test?

**Review by johnrandolph — COMMENTED** [review]
This is much improved formatting / organization. Just two questions and also are we going to generate coverage reports for this?

File: udmis/src/main/java/com/google/bos/udmi/service/messaging/impl/MessageDispatcherImpl.java [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+import org.jetbrains.annotations.TestOnly;
+import udmi.schema.Envelope;
+import udmi.schema.Envelope.SubFolder;
+import udmi.schema.Envelope.SubType;
+import udmi.schema.SystemState;
+
+/**
+ * Implementation of the typed message dispatcher interface.
+ */
+public class MessageDispatcherImpl extends ContainerBase implements MessageDispatcher {
```
[the remark this answers is not in the document]
Comment (grafnu) [reply udmis/src/main/java/com/google/bos/udmi/service/messaging/impl/MessageDispatcherImpl.java:36]: It's the only implementation of the interface. This is a common Java pattern. Java has no concept of a "header file" so  the separation likely serves some similar purpose.

File: udmis/src/main/java/com/google/bos/udmi/service/messaging/impl/MessageDispatcherImpl.java (hunk 1 above)
Comment (grafnu) [reply udmis/src/main/java/com/google/bos/udmi/service/messaging/impl/MessageDispatcherImpl.java:181]: Done

**Comment by grafnu** [comment]
Updated, PTAL.

**Review by johnrandolph — APPROVED** [review]

**Pull request merged**